### PR TITLE
Rewrite IPlaceholderOperation to IConditionalAccessInstanceOperation …

### PR DESCRIPTION
**Pull request by 333fred**
…when in that context. Fixes https://github.com/dotnet/roslyn/issues/21671.
vso: https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/521709

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 1]
```diff
@@ -939,12 +939,21 @@ Namespace Microsoft.CodeAnalysis.Operations
...
         End Function
 
-        Private Function CreateBoundRValuePlaceholderOperation(boundRValuePlaceholder As BoundRValuePlaceholder) As IPlaceholderOperation
+        Private Function CreateBoundRValuePlaceholderOperation(boundRValuePlaceholder As BoundRValuePlaceholder) As IOperation
             Dim syntax As SyntaxNode = boundRValuePlaceholder.Syntax
             Dim type As ITypeSymbol = boundRValuePlaceholder.Type
             Dim constantValue As [Optional](Of Object) = ConvertToOptional(boundRValuePlaceholder.ConstantValueOpt)
             Dim isImplicit As Boolean = boundRValuePlaceholder.WasCompilerGenerated
-            Return New PlaceholderExpression(_semanticModel, syntax, type, constantValue, isImplicit)
+            If syntax.IsKind(SyntaxKind.ConditionalAccessExpression) Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >If syntax.IsKind(SyntaxKind.ConditionalAccessExpression) Then [](start = 12, length = 61)

The syntax check is fine for now, but we should come up with a more robust way of detecting conditional access receivers. Either keep the original issue open, or open a new one.

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: For example, we could try creating node for the placeholder when we handle BoundConditionalAccess in CreateBoundConditionalAccessOperation. The operation will be cached and this method will never be called for this bound node.

---
In reply to: [149583274](https://github.com/dotnet/roslyn/pull/23068#discussion_r149583274) [](ancestors = 149583274)

**Review by AlekseyTs — APPROVED** [review]
LGTM as a temporary solution to meet the deadline. However, see if you have time to try the suggested alternative approach.

**Comment by AlekseyTs** [comment]
Just looked at implementation of VisualBasicOperationFactory.Create, it looks like it bypasses the cache for placeholder nodes as a workaround for other issues. So, the suggested alternative approach won't work right now.

**Comment by 333fred** [comment]
@cston could I get a second review from compilers?

**Review by mavasani — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 2]
```diff
@@ -939,12 +939,21 @@ Namespace Microsoft.CodeAnalysis.Operations
...
+                ' BoundConditionalAccessReceiver isn't actually used until local rewriting, until then that node will be a
+                ' BoundRValuePlaceholder with a sytnax node of the entire conditional access. So we dig through the syntax
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Typo: `syntax`

File: src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb [hunk 3]
```diff
@@ -1891,13 +1891,21 @@ End Class
...
                 Diagnostic(ConditionalAccessOperationTestAnalyzer.ConditionalAccessOperationDescriptor.Id, "Field1?.Field").WithLocation(30, 13),
+                Diagnostic(ConditionalAccessOperationTestAnalyzer.ConditionalAccessInstanceOperationDescriptor.Id, "Field1").WithLocation(30, 13),
                 Diagnostic(ConditionalAccessOperationTestAnalyzer.ConditionalAccessOperationDescriptor.Id, "Field1?(0)").WithLocation(31, 13),
-                Diagnostic(ConditionalAccessOperationTestAnalyzer.ConditionalAccessOperationDescriptor.Id, "Field1?.M0(Nothing)").WithLocation(32, 9))
+                Diagnostic(ConditionalAccessOperationTestAnalyzer.ConditionalAccessInstanceOperationDescriptor.Id, "Field1").WithLocation(31, 13),
+                Diagnostic(ConditionalAccessOperationTestAnalyzer.ConditionalAccessOperationDescriptor.Id, "Field1?.M0(Nothing)").WithLocation(32, 9),
+                Diagnostic(ConditionalAccessOperationTestAnalyzer.ConditionalAccessInstanceOperationDescriptor.Id, "Field1").WithLocation(32, 9))
         End Sub
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb]: Was this test failing previously?

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb (hunk 3 above)
Comment (333fred) [reply src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb]: It was not failing, but it was missing all the instance diagnostics.

**Comment by jinujoseph** [comment]
@natidea @tmat  this is good to merge once shiproom approves

**Comment by jinujoseph** [comment]
approved to merge via [Link](https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/521709)

**Pull request merged**